File: java/exercises/src/main/java/Webshop/Service/Users/UserService.java
package Webshop.Service.Users;

import java.util.UUID;
import Webshop.Service.Common.PasswordValidator;

interface IUserService {
    User createUser(String email, String firstName, String lastName, String password) throws InvalidPasswordException;
    User getUser(String userId);
}

public class UserService implements IUserService {
    private final IUserRepository userRepository;
    private final PasswordValidator passwordValidator;

    public UserService(UserRepository userRepository, PasswordValidator passwordValidator) {
        this.userRepository = userRepository;
        this.passwordValidator = passwordValidator;
    }

    @Override
    public User createUser(String email, String firstName, String lastName, String password) throws InvalidPasswordException {
        if (!passwordValidator.isValidPassword(password)) {
            throw new InvalidPasswordException("Password does not meet requirements");
        }

        String id = UUID.randomUUID().toString();

        return userRepository.addUser(id, email, firstName, lastName, password);
    }

    @Override
    public User getUser(String userId) {
        return userRepository.getUser(userId);
    }
}

class InvalidPasswordException extends Exception {
    public InvalidPasswordException(String message) {
        super(message);
    }
}

File: csharp/exercises/Webshop.Service/Users/UserService.cs
﻿using Webshop.Service.Common;

namespace Webshop.Service.Users;

public interface IUserService
{
    User CreateUser(string email, string firstName, string lastName, string password);
    User? GetUser(string userId);
}

public class UserService : IUserService
{
    private readonly IUserRepository _userRepository;
    private readonly IPasswordValidator _passwordValidator;

    public UserService(IUserRepository userRepository, IPasswordValidator passwordValidator)
    {
        _userRepository = userRepository;
        _passwordValidator = passwordValidator;
    }

    public User CreateUser(string email, string firstName, string lastName, string password)
    {
        if (!_passwordValidator.IsValidPassword(password))
            throw new InvalidPasswordException("Password does not meet requirements");

        var id = Guid.NewGuid().ToString();

        return _userRepository.AddUser(id, email, firstName, lastName, password);
    }

    public User? GetUser(string userId)
    {
        return _userRepository.GetUser(userId);
    }
}

public class InvalidPasswordException : Exception
{
    public InvalidPasswordException(string message) : base(message) { }
}